File: pkg/go_grim/grim_reaper.go
// Package grim provides real core integration with sh_grim, py_grim, and go_grim
// No mock files - directly calls actual core modules and binaries
package grim

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"time"
)

// GrimReaper provides integration with Grim Reaper core components
type GrimReaper struct {
	GrimRoot string
	APIBase  string
}

// BackupOptions represents backup configuration
type BackupOptions struct {
	Name        string
	Compress    string
	Incremental bool
}

// CompressionOptions represents compression configuration
type CompressionOptions struct {
	Algorithm string
	Level     int
	Output    string
}

// MonitoringOptions represents monitoring configuration
type MonitoringOptions struct {
	Interval int
	Events   string
}

// ScanOptions represents scanning configuration
type ScanOptions struct {
	Recursive bool
	Types     string
	Output    string
}

// SecurityScanOptions represents security scan configuration
type SecurityScanOptions struct {
	Deep   bool
	Report string
}

// NewGrimReaper creates a new GrimReaper instance with portable path discovery
func NewGrimReaper(grimRoot ...string) (*GrimReaper, error) {
	var root string
	var err error

	if len(grimRoot) > 0 && grimRoot[0] != "" {
		root = grimRoot[0]
	} else {
		root, err = findGrimRoot()
		if err != nil {
			return nil, err
		}
	}

	return &GrimReaper{
		GrimRoot: root,
		APIBase:  "http://localhost:8000",
	}, nil
}

// findGrimRoot searches for Grim Reaper installation directory
func findGrimRoot() (string, error) {
	// Check environment variable first
	if envPath := os.Getenv("GRIM_ROOT"); envPath != "" {
		if isGrimInstallation(envPath) {
			return envPath, nil
		}
	}

	// Search up directory tree
	currentDir, err := os.Getwd()
	if err != nil {
		return "", err
	}

	for i := 0; i < 10; i++ {
		if isGrimInstallation(currentDir) {
			return currentDir, nil
		}

		parentDir := filepath.Dir(currentDir)
		if parentDir == currentDir {
			break
		}
		currentDir = parentDir
	}

	// Try common installation paths
	possiblePaths := []string{
		filepath.Join(os.Getenv("HOME"), "reaper"),
		filepath.Join(os.Getenv("HOME"), ".reaper"),
		"/root/reaper",
		"/root/.reaper", 
		"/usr/local/reaper",
		"/usr/local/share/reaper",
		"/usr/share/reaper",
		"/opt/reaper",
		"/usr/local/lib/grim-reaper",
		"/usr/lib/grim-reaper",
	}

	for _, path := range possiblePaths {
		if isGrimInstallation(path) {
			return path, nil
		}
	}

	// Grim Reaper not found - attempt automatic installation
	fmt.Println("Grim Reaper not found. Attempting automatic installation...")
	installPath, err := installGrimReaper()
	if err != nil {
		return "", fmt.Errorf(`could not find Grim Reaper root directory and automatic installation failed: %v

Please ensure Grim Reaper is properly installed using:
  • curl -fsSL https://get.grim.so | sudo bash
  • wget -qO- https://get.grim.so | sudo bash

Or set GRIM_ROOT environment variable:
  export GRIM_ROOT=/path/to/your/grim/installation`, err)
	}

	fmt.Printf("Grim Reaper automatically installed to: %s\n", installPath)
	return installPath, nil
}

// installGrimReaper automatically installs Grim Reaper
func installGrimReaper() (string, error) {
	// Determine installation directory
	var installDir string
	if os.Geteuid() == 0 {
		// Running as root - install system-wide
		installDir = "/opt/reaper"
	} else {
		// Running as user - install to home directory
		homeDir, err := os.UserHomeDir()
		if err != nil {
			return "", fmt.Errorf("failed to get home directory: %v", err)
		}
		installDir = filepath.Join(homeDir, ".reaper")
	}

	// Create temporary directory for download
	tempDir, err := os.MkdirTemp("", "grim-install-*")
	if err != nil {
		return "", fmt.Errorf("failed to create temp directory: %v", err)
	}
	defer os.RemoveAll(tempDir)

	// Download installer
	installerPath := filepath.Join(tempDir, "install.sh")
	if err := downloadFile("https://get.grim.so", installerPath); err != nil {
		return "", fmt.Errorf("failed to download installer: %v", err)
	}

	// Make installer executable
	if err := os.Chmod(installerPath, 0755); err != nil {
		return "", fmt.Errorf("failed to make installer executable: %v", err)
	}

	// Run installer
	cmd := exec.Command(installerPath)
	cmd.Dir = tempDir
	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stderr
	
	if err := cmd.Run(); err != nil {
		return "", fmt.Errorf("installer failed: %v", err)
	}

	// Verify installation
	if isGrimInstallation(installDir) {
		return installDir, nil
	}

	// Try alternative paths
	altPaths := []string{
		"/root/.graveyard/reaper",
		filepath.Join(os.Getenv("HOME"), ".graveyard/reaper"),
	}

	for _, path := range altPaths {
		if isGrimInstallation(path) {
			return path, nil
		}
	}

	return "", fmt.Errorf("installation completed but Grim Reaper not found in expected locations")
}

// downloadFile downloads a file from URL to local path
func downloadFile(url, filepath string) error {
	resp, err := http.Get(url)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return fmt.Errorf("bad status: %s", resp.Status)
	}

	out, err := os.Create(filepath)
	if err != nil {
		return err
	}
	defer out.Close()

	_, err = io.Copy(out, resp.Body)
	return err
}

// isGrimInstallation checks if path contains a valid Grim installation
func isGrimInstallation(path string) bool {
	if _, err := os.Stat(path); os.IsNotExist(err) {
		return false
	}

	// Check for key Grim files
	keyFiles := []string{
		"throne/grim_throne.sh",
		"tsk_flask/grim_admin_server.py",
		"sh_grim/backup.sh",
		"go_grim/build/grim-compression",
	}

	for _, keyFile := range keyFiles {
		if _, err := os.Stat(filepath.Join(path, keyFile)); err == nil {
			return true
		}
	}

	return false
}

// executeShModule executes sh_grim module with proper error handling
func (g *GrimReaper) executeShModule(module string, args ...string) (string, error) {
	modulePath := filepath.Join(g.GrimRoot, "sh_grim", module+".sh")
	
	if _, err := os.Stat(modulePath); os.IsNotExist(err) {
		return "", fmt.Errorf("module not found: %s", module)
	}

	cmd := exec.Command(modulePath, args...)
	cmd.Dir = g.GrimRoot
	
	output, err := cmd.CombinedOutput()
	if err != nil {
		return "", fmt.Errorf("module %s failed: %v - %s", module, err, string(output))
	}

	return string(output), nil
}

// executeGoBinary executes go_grim binary with proper error handling
func (g *GrimReaper) executeGoBinary(binary string, args ...string) (string, error) {
	binaryPath := filepath.Join(g.GrimRoot, "go_grim", "build", binary)
	
	if _, err := os.Stat(binaryPath); os.IsNotExist(err) {
		return "", fmt.Errorf("go binary not found: %s", binary)
	}

	cmd := exec.Command(binaryPath, args...)
	cmd.Dir = g.GrimRoot
	
	output, err := cmd.CombinedOutput()
	if err != nil {
		return "", fmt.Errorf("go binary %s failed: %v - %s", binary, err, string(output))
	}

	return string(output), nil
}

// callPyAPI calls py_grim FastAPI service
func (g *GrimReaper) callPyAPI(endpoint string) (map[string]interface{}, error) {
	url := g.APIBase + endpoint
	
	client := &http.Client{Timeout: 30 * time.Second}
	resp, err := client.Get(url)
	if err != nil {
		return nil, fmt.Errorf("API call failed: %v", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("API call failed with status: %d", resp.StatusCode)
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("failed to read API response: %v", err)
	}

	var result map[string]interface{}
	if err := json.Unmarshal(body, &result); err != nil {
		return nil, fmt.Errorf("failed to parse API response: %v", err)
	}

	return result, nil
}

// ============================================================================
// BACKUP OPERATIONS (via sh_grim)
// ============================================================================

// Backup creates backup using sh_grim/backup.sh
func (g *GrimReaper) Backup(source string, opts *BackupOptions) (string, error) {
	args := []string{source}
	
	if opts != nil {
		if opts.Name != "" {
			args = append(args, "--name", opts.Name)
		}
		if opts.Compress != "" {
			args = append(args, "--compress", opts.Compress)
		}
		if opts.Incremental {
			args = append(args, "--incremental")
		}
	}

	return g.executeShModule("backup", args...)
}

// Restore restores from backup using sh_grim/restore.sh
func (g *GrimReaper) Restore(backup, destination string, overwrite bool) (string, error) {
	args := []string{backup, destination}
	
	if overwrite {
		args = append(args, "--overwrite")
	}

	return g.executeShModule("restore", args...)
}

// ListBackups lists available backups
func (g *GrimReaper) ListBackups() (string, error) {
	return g.executeShModule("backup", "--list")
}

// ============================================================================
// COMPRESSION OPERATIONS (via go_grim)
// ============================================================================

// Compress compresses file using go_grim compression engine
func (g *GrimReaper) Compress(filePath string, opts *CompressionOptions) (string, error) {
	args := []string{}
	
	if opts != nil {
		if opts.Algorithm != "" {
			args = append(args, "-a", opts.Algorithm)
		}
		if opts.Level > 0 {
			args = append(args, "-l", fmt.Sprintf("%d", opts.Level))
		}
		if opts.Output != "" {
			args = append(args, "-o", opts.Output)
		}
	}
	
	args = append(args, filePath)

	return g.executeGoBinary("grim-compression", args...)
}

// Decompress decompresses file using go_grim
func (g *GrimReaper) Decompress(filePath, output string) (string, error) {
	args := []string{"-d"}
	
	if output != "" {
		args = append(args, "-o", output)
	}
	args = append(args, filePath)

	return g.executeGoBinary("grim-compression", args...)
}

// BenchmarkCompression gets compression benchmarks
func (g *GrimReaper) BenchmarkCompression(filePath string) (string, error) {
	return g.executeGoBinary("grim-compression", "-benchmark", filePath)
}

// ============================================================================
// MONITORING OPERATIONS (via sh_grim)
// ============================================================================

// StartMonitoring starts monitoring using sh_grim/monitor.sh
func (g *GrimReaper) StartMonitoring(path string, opts *MonitoringOptions) (string, error) {
	args := []string{"start", path}
	
	if opts != nil {
		if opts.Interval > 0 {
			args = append(args, "--interval", fmt.Sprintf("%d", opts.Interval))
		}
		if opts.Events != "" {
			args = append(args, "--events", opts.Events)
		}
	}

	return g.executeShModule("monitor", args...)
}

// StopMonitoring stops monitoring
func (g *GrimReaper) StopMonitoring() (string, error) {
	return g.executeShModule("monitor", "stop")
}

// GetMonitoringStatus gets monitoring status
func (g *GrimReaper) GetMonitoringStatus() (string, error) {
	return g.executeShModule("monitor", "status")
}

// ============================================================================
// SCANNING OPERATIONS (via sh_grim)
// ============================================================================

// Scan scans directory using sh_grim/scan.sh
func (g *GrimReaper) Scan(path string, opts *ScanOptions) (string, error) {
	args := []string{path}
	
	if opts != nil {
		if opts.Recursive {
			args = append(args, "--recursive")
		}
		if opts.Types != "" {
			args = append(args, "--types", opts.Types)
		}
		if opts.Output != "" {
			args = append(args, "--output", opts.Output)
		}
	}

	return g.executeShModule("scan", args...)
}

// SecurityScan performs security scan using sh_grim/security.sh
func (g *GrimReaper) SecurityScan(path string, opts *SecurityScanOptions) (string, error) {
	args := []string{path}
	
	if opts != nil {
		if opts.Deep {
			args = append(args, "--deep")
		}
		if opts.Report != "" {
			args = append(args, "--report", opts.Report)
		}
	}

	return g.executeShModule("security", args...)
}

// ============================================================================
// SYSTEM OPERATIONS (via sh_grim)
// ============================================================================

// HealthCheck performs system health check using sh_grim/health.sh
func (g *GrimReaper) HealthCheck() (string, error) {
	return g.executeShModule("health", "check")
}

// GetStatus gets system status
func (g *GrimReaper) GetStatus() (string, error) {
	return g.executeShModule("health", "status")
}

// Optimize optimizes system using sh_grim/blacksmith.sh
func (g *GrimReaper) Optimize(target string) (string, error) {
	if target == "" {
		target = "all"
	}
	return g.executeShModule("blacksmith", "optimize", target)
}

// Heal performs self-healing using sh_grim/healer.sh
func (g *GrimReaper) Heal() (string, error) {
	return g.executeShModule("healer", "heal")
}

// ============================================================================
// API INTEGRATION (via py_grim FastAPI)
// ============================================================================

// GetAPIStatus gets system status via API
func (g *GrimReaper) GetAPIStatus() (map[string]interface{}, error) {
	return g.callPyAPI("/api/status")
}

// GetBackupInfo gets backup information via API
func (g *GrimReaper) GetBackupInfo() (map[string]interface{}, error) {
	return g.callPyAPI("/api/backups")
}

// GetMonitoringData gets monitoring data via API
func (g *GrimReaper) GetMonitoringData() (map[string]interface{}, error) {
	return g.callPyAPI("/api/monitoring")
}

// ============================================================================
// UTILITY METHODS
// ============================================================================

// ExecuteCommand executes raw grim command via throne script
func (g *GrimReaper) ExecuteCommand(command string, args ...string) (string, error) {
	thronePath := filepath.Join(g.GrimRoot, "throne", "grim_throne.sh")
	
	if _, err := os.Stat(thronePath); os.IsNotExist(err) {
		return "", fmt.Errorf("throne script not found: %s", thronePath)
	}

	cmdArgs := append([]string{command}, args...)
	cmd := exec.Command(thronePath, cmdArgs...)
	cmd.Dir = g.GrimRoot
	
	output, err := cmd.CombinedOutput()
	if err != nil {
		return "", fmt.Errorf("command %s failed: %v - %s", command, err, string(output))
	}

	return string(output), nil
}

// GetVersion gets Grim version and build info
func (g *GrimReaper) GetVersion() (string, error) {
	manifestPath := filepath.Join(g.GrimRoot, "builds", "latest", "manifest.tsk")
	
	if _, err := os.Stat(manifestPath); err == nil {
		content, err := os.ReadFile(manifestPath)
		if err == nil {
			return string(content), nil
		}
	}

	return g.ExecuteCommand("version")
}

// CheckServices checks if Grim services are running
func (g *GrimReaper) CheckServices() map[string]bool {
	services := map[string]bool{
		"api":        false,
		"monitoring": false,
		"admin":      false,
	}

	// Check FastAPI service
	if cmd := exec.Command("pgrep", "-f", "grim_web"); cmd.Run() == nil {
		services["api"] = true
	}

	// Check monitoring service
	if cmd := exec.Command("pgrep", "-f", "monitor.sh"); cmd.Run() == nil {
		services["monitoring"] = true
	}

	// Check admin server
	if cmd := exec.Command("pgrep", "-f", "grim_admin_server.py"); cmd.Run() == nil {
		services["admin"] = true
	}

	return services
}

// SetAPIBase sets the API base URL
func (g *GrimReaper) SetAPIBase(apiBase string) {
	g.APIBase = strings.TrimSuffix(apiBase, "/")
}

// GetGrimRoot returns the Grim root directory
func (g *GrimReaper) GetGrimRoot() string {
	return g.GrimRoot
}

// GetAPIBase returns the API base URL
func (g *GrimReaper) GetAPIBase() string {
	return g.APIBase
}

// ============================================================================
// CONVENIENCE FUNCTIONS
// ============================================================================

// InstallGrimReaper installs Grim Reaper if not already installed
func InstallGrimReaper() (string, error) {
	// First check if already installed
	if root, err := findGrimRoot(); err == nil {
		return root, nil
	}
	
	// Not installed - install it
	return installGrimReaper()
}

// NewGrimReaperWithInstall creates a new GrimReaper instance, installing if needed
func NewGrimReaperWithInstall(grimRoot ...string) (*GrimReaper, error) {
	var root string
	var err error

	if len(grimRoot) > 0 && grimRoot[0] != "" {
		root = grimRoot[0]
	} else {
		root, err = findGrimRoot() // This now includes auto-installation
		if err != nil {
			return nil, err
		}
	}

	return &GrimReaper{
		GrimRoot: root,
		APIBase:  "http://localhost:8000",
	}, nil
}

// QuickBackup performs a backup with automatic installation if needed
func QuickBackup(source string) (string, error) {
	grim, err := NewGrimReaperWithInstall()
	if err != nil {
		return "", err
	}
	return grim.Backup(source, nil)
}

// QuickRestore performs a restore with automatic installation if needed
func QuickRestore(backup, destination string) (string, error) {
	grim, err := NewGrimReaperWithInstall()
	if err != nil {
		return "", err
	}
	return grim.Restore(backup, destination, false)
}

// QuickCompress compresses a file with automatic installation if needed
func QuickCompress(filePath string) (string, error) {
	grim, err := NewGrimReaperWithInstall()
	if err != nil {
		return "", err
	}
	return grim.Compress(filePath, nil)
}

// QuickHealthCheck performs a health check with automatic installation if needed
func QuickHealthCheck() (string, error) {
	grim, err := NewGrimReaperWithInstall()
	if err != nil {
		return "", err
	}
	return grim.HealthCheck()
}

// QuickScan performs a scan with automatic installation if needed
func QuickScan(path string) (string, error) {
	grim, err := NewGrimReaperWithInstall()
	if err != nil {
		return "", err
	}
	return grim.Scan(path, nil)
}

File: pkg/php_grim/src/GrimReaper.php
<?php

/**
 * Grim Reaper PHP Package
 * Real core integration with sh_grim, py_grim, and go_grim
 * No mock files - calls actual core modules and binaries
 *
 * @copyright 2025 Bernie Gengel and CyberBoost LLC
 * @license MIT License - see LICENSE file for full terms
 * @package GrimReaper
 */

namespace GrimReaper;
class GrimReaper
{
    private string $grimRoot;
    private string $apiBase;

    public function __construct(?string $grimRoot = null)
    {
        $this->grimRoot = $grimRoot ?: $this->findGrimRoot();
        $this->apiBase = 'http://localhost:8000';
    }

    /**
     * Find Grim Reaper installation directory using portable discovery
     */
    private function findGrimRoot(): string
    {
        // Check environment variable first
        $envPath = $_ENV['GRIM_ROOT'] ?? $_SERVER['GRIM_ROOT'] ?? null;
        if ($envPath && $this->isGrimInstallation($envPath)) {
            return $envPath;
        }

        // Search up directory tree
        $currentDir = getcwd();
        for ($i = 0; $i < 10; $i++) {
            if ($this->isGrimInstallation($currentDir)) {
                return $currentDir;
            }
            
            $parentDir = dirname($currentDir);
            if ($parentDir === $currentDir) {
                break;
            }
            $currentDir = $parentDir;
        }

        // Try common installation paths
        $possiblePaths = [
            $_SERVER['HOME'] . '/reaper',
            $_SERVER['HOME'] . '/.reaper',
            '/root/reaper',
            '/root/.reaper',
            '/usr/local/reaper',
            '/usr/local/share/reaper',
            '/usr/share/reaper',
            '/opt/reaper',
            '/usr/local/lib/grim-reaper',
            '/usr/lib/grim-reaper',
        ];

        foreach ($possiblePaths as $path) {
            if ($this->isGrimInstallation($path)) {
                return $path;
            }
        }

        throw new \RuntimeException(
            "Could not find Grim Reaper root directory.\n" .
            "Please ensure Grim Reaper is properly installed using:\n" .
            "  • curl -fsSL https://get.grim.so | sudo bash\n" .
            "  • wget -qO- https://get.grim.so | sudo bash\n\n" .
            "Or set GRIM_ROOT environment variable:\n" .
            "  export GRIM_ROOT=/path/to/your/grim/installation"
        );
    }

    /**
     * Check if path contains a valid Grim installation
     */
    private function isGrimInstallation(string $path): bool
    {
        if (!is_dir($path)) {
            return false;
        }

        // Check for key Grim files
        $keyFiles = [
            'throne/grim_throne.sh',
            'tsk_flask/grim_admin_server.py',
            'sh_grim/backup.sh',
            'go_grim/build/grim-compression',
        ];

        foreach ($keyFiles as $keyFile) {
            if (file_exists($path . '/' . $keyFile)) {
                return true;
            }
        }

        return false;
    }

    /**
     * Execute sh_grim module with proper error handling
     */
    private function executeShModule(string $module, array $args = []): string
    {
        $modulePath = $this->grimRoot . "/sh_grim/{$module}.sh";
        
        if (!file_exists($modulePath)) {
            throw new \RuntimeException("Module not found: {$module}");
        }

        if (!is_executable($modulePath)) {
            chmod($modulePath, 0755);
        }

        $command = escapeshellcmd($modulePath);
        if (!empty($args)) {
            $command .= ' ' . implode(' ', array_map('escapeshellarg', $args));
        }

        $output = [];
        $returnCode = 0;
        
        exec("cd {$this->grimRoot} && {$command} 2>&1", $output, $returnCode);

        if ($returnCode !== 0) {
            throw new \RuntimeException("Module {$module} failed: " . implode("\n", $output));
        }

        return implode("\n", $output);
    }

    /**
     * Execute go_grim binary with proper error handling
     */
    private function executeGoBinary(string $binary, array $args = []): string
    {
        $binaryPath = $this->grimRoot . "/go_grim/build/{$binary}";
        
        if (!file_exists($binaryPath)) {
            throw new \RuntimeException("Go binary not found: {$binary}");
        }

        if (!is_executable($binaryPath)) {
            chmod($binaryPath, 0755);
        }

        $command = escapeshellcmd($binaryPath);
        if (!empty($args)) {
            $command .= ' ' . implode(' ', array_map('escapeshellarg', $args));
        }

        $output = [];
        $returnCode = 0;
        
        exec("cd {$this->grimRoot} && {$command} 2>&1", $output, $returnCode);

        if ($returnCode !== 0) {
            throw new \RuntimeException("Go binary {$binary} failed: " . implode("\n", $output));
        }

        return implode("\n", $output);
    }

    /**
     * Call py_grim FastAPI service
     */
    private function callPyApi(string $endpoint, string $method = 'GET', ?array $data = null): array
    {
        $url = $this->apiBase . $endpoint;
        
        $context = stream_context_create([
            'http' => [
                'method' => $method,
                'header' => 'Content-Type: application/json',
                'timeout' => 30,
            ]
        ]);

        if ($method === 'POST' && $data) {
            $context['http']['content'] = json_encode($data);
        }

        $response = @file_get_contents($url, false, $context);
        
        if ($response === false) {
            throw new \RuntimeException("API call failed: {$endpoint}");
        }

        $json = json_decode($response, true);
        if (json_last_error() !== JSON_ERROR_NONE) {
            throw new \RuntimeException("Invalid JSON response from API");
        }

        return $json;
    }

    // ============================================================================
    // BACKUP OPERATIONS (via sh_grim)
    // ============================================================================

    /**
     * Create backup using sh_grim/backup.sh
     */
    public function backup(string $source, ?string $name = null, string $compress = 'zstd', bool $incremental = false): string
    {
        $args = [$source];
        
        if ($name) {
            $args[] = '--name';
            $args[] = $name;
        }
        
        if ($compress) {
            $args[] = '--compress';
            $args[] = $compress;
        }
        
        if ($incremental) {
            $args[] = '--incremental';
        }

        return $this->executeShModule('backup', $args);
    }

    /**
     * Restore from backup using sh_grim/restore.sh
     */
    public function restore(string $backup, string $destination, bool $overwrite = false): string
    {
        $args = [$backup, $destination];
        
        if ($overwrite) {
            $args[] = '--overwrite';
        }

        return $this->executeShModule('restore', $args);
    }

    /**
     * List available backups
     */
    public function listBackups(): string
    {
        return $this->executeShModule('backup', ['--list']);
    }

    // ============================================================================
    // COMPRESSION OPERATIONS (via go_grim)
    // ============================================================================

    /**
     * Compress file using go_grim compression engine
     */
    public function compress(string $filePath, string $algorithm = 'zstd', int $level = 6, ?string $output = null): string
    {
        $args = [];
        
        if ($algorithm) {
            $args[] = '-a';
            $args[] = $algorithm;
        }
        
        if ($level) {
            $args[] = '-l';
            $args[] = (string)$level;
        }
        
        if ($output) {
            $args[] = '-o';
            $args[] = $output;
        }
        
        $args[] = $filePath;

        return $this->executeGoBinary('grim-compression', $args);
    }

    /**
     * Decompress file using go_grim
     */
    public function decompress(string $filePath, ?string $output = null): string
    {
        $args = ['-d'];
        
        if ($output) {
            $args[] = '-o';
            $args[] = $output;
        }
        
        $args[] = $filePath;

        return $this->executeGoBinary('grim-compression', $args);
    }

    /**
     * Get compression benchmarks
     */
    public function benchmarkCompression(string $filePath): string
    {
        return $this->executeGoBinary('grim-compression', ['-benchmark', $filePath]);
    }

    // ============================================================================
    // MONITORING OPERATIONS (via sh_grim)
    // ============================================================================

    /**
     * Start monitoring using sh_grim/monitor.sh
     */
    public function startMonitoring(string $path, int $interval = 5, string $events = 'all'): string
    {
        $args = ['start', $path];
        
        if ($interval) {
            $args[] = '--interval';
            $args[] = (string)$interval;
        }
        
        if ($events) {
            $args[] = '--events';
            $args[] = $events;
        }

        return $this->executeShModule('monitor', $args);
    }

    /**
     * Stop monitoring
     */
    public function stopMonitoring(): string
    {
        return $this->executeShModule('monitor', ['stop']);
    }

    /**
     * Get monitoring status
     */
    public function getMonitoringStatus(): string
    {
        return $this->executeShModule('monitor', ['status']);
    }

    // ============================================================================
    // SCANNING OPERATIONS (via sh_grim)
    // ============================================================================

    /**
     * Scan directory using sh_grim/scan.sh
     */
    public function scan(string $path, bool $recursive = true, ?string $types = null, ?string $output = null): string
    {
        $args = [$path];
        
        if ($recursive) {
            $args[] = '--recursive';
        }
        
        if ($types) {
            $args[] = '--types';
            $args[] = $types;
        }
        
        if ($output) {
            $args[] = '--output';
            $args[] = $output;
        }

        return $this->executeShModule('scan', $args);
    }

    /**
     * Security scan using sh_grim/security.sh
     */
    public function securityScan(string $path, bool $deep = false, ?string $report = null): string
    {
        $args = [$path];
        
        if ($deep) {
            $args[] = '--deep';
        }
        
        if ($report) {
            $args[] = '--report';
            $args[] = $report;
        }

        return $this->executeShModule('security', $args);
    }

    // ============================================================================
    // SYSTEM OPERATIONS (via sh_grim) 
    // ============================================================================

    /**
     * System health check using sh_grim/health.sh
     */
    public function healthCheck(): string
    {
        return $this->executeShModule('health', ['check']);
    }

    /**
     * Get system status
     */
    public function getStatus(): string
    {
        return $this->executeShModule('health', ['status']);
    }

    /**
     * Optimize system using sh_grim/blacksmith.sh
     */
    public function optimize(string $target = 'all'): string
    {
        return $this->executeShModule('blacksmith', ['optimize', $target]);
    }

    /**
     * Self-healing using sh_grim/healer.sh
     */
    public function heal(): string
    {
        return $this->executeShModule('healer', ['heal']);
    }

    // ============================================================================
    // API INTEGRATION (via py_grim FastAPI)
    // ============================================================================

    /**
     * Get system status via API
     */
    public function getApiStatus(): array
    {
        return $this->callPyApi('/api/status');
    }

    /**
     * Get backup information via API
     */
    public function getBackupInfo(): array
    {
        return $this->callPyApi('/api/backups');
    }

    /**
     * Start backup via API
     */
    public function startApiBackup(string $source, array $options = []): array
    {
        $data = array_merge(['source' => $source], $options);
        return $this->callPyApi('/api/backup', 'POST', $data);
    }

    /**
     * Get monitoring data via API
     */
    public function getMonitoringData(): array
    {
        return $this->callPyApi('/api/monitoring');
    }

    // ============================================================================
    // UTILITY METHODS
    // ============================================================================

    /**
     * Execute raw grim command via throne script
     */
    public function executeCommand(string $command, array $args = []): string
    {
        $thronePath = $this->grimRoot . '/throne/grim_throne.sh';
        
        if (!file_exists($thronePath)) {
            throw new \RuntimeException("Throne script not found: {$thronePath}");
        }

        if (!is_executable($thronePath)) {
            chmod($thronePath, 0755);
        }

        $fullCommand = escapeshellcmd($thronePath) . ' ' . escapeshellarg($command);
        if (!empty($args)) {
            $fullCommand .= ' ' . implode(' ', array_map('escapeshellarg', $args));
        }

        $output = [];
        $returnCode = 0;
        
        exec("cd {$this->grimRoot} && {$fullCommand} 2>&1", $output, $returnCode);

        if ($returnCode !== 0) {
            throw new \RuntimeException("Command {$command} failed: " . implode("\n", $output));
        }

        return implode("\n", $output);
    }

    /**
     * Get Grim version and build info
     */
    public function getVersion(): string
    {
        $manifestPath = $this->grimRoot . '/builds/latest/manifest.tsk';
        
        if (file_exists($manifestPath)) {
            return file_get_contents($manifestPath) ?: '';
        }

        return $this->executeCommand('version');
    }

    /**
     * Check if Grim services are running
     */
    public function checkServices(): array
    {
        $services = [
            'api' => false,
            'monitoring' => false,
            'admin' => false,
        ];

        // Check FastAPI service
        exec('pgrep -f grim_web', $output, $returnCode);
        $services['api'] = ($returnCode === 0);

        // Check monitoring service
        exec('pgrep -f monitor.sh', $output, $returnCode);
        $services['monitoring'] = ($returnCode === 0);

        // Check admin server
        exec('pgrep -f grim_admin_server.py', $output, $returnCode);
        $services['admin'] = ($returnCode === 0);

        return $services;
    }

    /**
     * Get Grim root directory
     */
    public function getGrimRoot(): string
    {
        return $this->grimRoot;
    }

    /**
     * Get API base URL
     */
    public function getApiBase(): string
    {
        return $this->apiBase;
    }

    /**
     * Set API base URL
     */
    public function setApiBase(string $apiBase): void
    {
        $this->apiBase = $apiBase;
    }
}

// Convenience functions for direct use
function grim_backup(string $source, array $options = []): string
{
    $grim = new GrimReaper();
    return $grim->backup($source, $options['name'] ?? null, $options['compress'] ?? 'zstd', $options['incremental'] ?? false);
}

function grim_restore(string $backup, string $destination, bool $overwrite = false): string
{
    $grim = new GrimReaper();
    return $grim->restore($backup, $destination, $overwrite);
}

function grim_compress(string $filePath, array $options = []): string
{
    $grim = new GrimReaper();
    return $grim->compress($filePath, $options['algorithm'] ?? 'zstd', $options['level'] ?? 6, $options['output'] ?? null);
}

function grim_health_check(): string
{
    $grim = new GrimReaper();
    return $grim->healthCheck();
}

function grim_scan(string $path, array $options = []): string
{
    $grim = new GrimReaper();
    return $grim->scan($path, $options['recursive'] ?? true, $options['types'] ?? null, $options['output'] ?? null);
}